pkg/natdiscovery/natdiscovery.go

```diff
@@ -1 +1,188 @@
+/*
+© 2021 Red Hat, Inc. and others.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
 package natdiscovery
+
+import (
+	"crypto/rand"
+	"math/big"
+	"net"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+
+	v1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	"github.com/submariner-io/submariner/pkg/types"
+)
+
+type Interface interface {
+	Run(stopCh <-chan struct{}, natResultCh chan remoteEndpointNAT) error
+	AddEndpoint(endpoint *types.SubmarinerEndpoint)
+	RemoveEndpoint(endpointName string)
+	GetEndpointNAT(endpointName string) (bool, error)
+	GetEndpointIP(endpointName string) (string, error)
+}
+
+type udpWriteFunction func(b []byte, addr *net.UDPAddr) (int, error)
+type findSrcIPFunction func(destinationIP string) (string, error)
+
+type natDiscovery struct {
+	sync.Mutex
+	localEndpoint   *types.SubmarinerEndpoint
+	remoteEndpoints map[string]*remoteEndpointNAT
+	requestCounter  uint64
+	serverUDPWrite  udpWriteFunction
+	findSrcIP       findSrcIPFunction
+	serverPort      int
+}
+
+func New(localEndpoint *types.SubmarinerEndpoint) (Interface, error) {
+	return newNatDiscovery(localEndpoint)
+}
+
+func newNatDiscovery(localEndpoint *types.SubmarinerEndpoint) (*natDiscovery, error) {
+	port, err := extractNATDiscoveryPort(localEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	requestCounter, err := randomRequestCounter()
+	if err != nil {
+		return nil, err
+	}
+
+	return &natDiscovery{
+		localEndpoint:   localEndpoint,
+		serverPort:      port,
+		remoteEndpoints: map[string]*remoteEndpointNAT{},
+		findSrcIP:       findPreferredSourceIP,
+		requestCounter:  requestCounter,
+	}, nil
+}
+
+func randomRequestCounter() (uint64, error) {
+	n, err := rand.Int(rand.Reader, big.NewInt(1000))
+	if err != nil {
+		return 0, errors.Wrapf(err, "generatimg random request counter")
+	}
+
+	return n.Uint64(), nil
+}
+
+var errorNoNatDiscoveryPort = errors.New("no nat discovery port in endpoint")
```

Comment:
nit: "nat discovery port missing in endpoint"

Reply:
I forgot that one, will do on next pass